src/workerd/api/node/buffer.c++

```diff
@@ -597,5 +606,211 @@ uint32_t BufferUtil::write(
   return writeInto(js, buffer, string, offset, length, getEncoding(encoding));
 }
 
+// ======================================================================================
+// StringDecoder
+
+namespace {
+inline kj::byte getMissingBytes(kj::ArrayPtr<kj::byte> state) {
+  return state[BufferUtil::kMissingBytes];
+}
+
+inline kj::byte getBufferedBytes(kj::ArrayPtr<kj::byte> state) {
+  return state[BufferUtil::kBufferedBytes];
+}
+
+inline kj::byte* getIncompleteCharacterBuffer(kj::ArrayPtr<kj::byte> state) {
+  return state.begin() + BufferUtil::kIncompleteCharactersStart;
+}
+
+inline Encoding getEncoding(kj::ArrayPtr<kj::byte> state) {
+  return static_cast<Encoding>(state[BufferUtil::kEncoding]);
+}
+
+v8::Local<v8::String> getBufferedString(jsg::Lock& js, kj::ArrayPtr<kj::byte> state) {
+  KJ_ASSERT(getBufferedBytes(state) <= BufferUtil::kIncompleteCharactersEnd);
+  auto ret = toStringImpl(js, state,
+                          BufferUtil::kIncompleteCharactersStart,
+                          BufferUtil::kIncompleteCharactersStart + getBufferedBytes(state),
+                          getEncoding(state));
+  state[BufferUtil::kBufferedBytes] = 0;
+  return ret;
+}
+}  // namespace
+
+v8::Local<v8::String> BufferUtil::decode(jsg::Lock& js,
+                                         kj::Array<kj::byte> bytes,
+                                         kj::Array<kj::byte> state) {
+  KJ_ASSERT(state.size() == BufferUtil::kSize);
+  auto enc = getEncoding(state);
+  if (enc == Encoding::ASCII || enc == Encoding::LATIN1 || enc == Encoding::HEX) {
+    // For ascii, latin1, and hex, we can just use the regular
+    // toString option since there will never be a case where
+    // these have left-over characters.
+    return toStringImpl(js, bytes, 0, bytes.size(), enc);
+  }
+
+  v8::Local<v8::String> prepend;
+  v8::Local<v8::String> body;
+  auto nread = bytes.size();
+  auto data = bytes.begin();
+  if (getMissingBytes(state) > 0) {
+    KJ_ASSERT(getMissingBytes(state) + getBufferedBytes(state) <=
+              BufferUtil::kIncompleteCharactersEnd);
+    if (enc == Encoding::UTF8) {
+      // For UTF-8, we need special treatment to algin with the V8 decoder:
+      // If an incomplete character is found at a chunk boundary, we use
+      // its remainder and pass it to V8 as-is.
+      for (size_t i = 0; i < nread && i < getMissingBytes(state); ++i) {
+        if ((bytes[i] & 0xC0) != 0x80) {
```

Comment:
there are a lot of buffer accesses here. Will we retain bounds check on runtime? Can we make sure we have it?

The code below is very dense and chances for buffer overruns are very real.

Maybe rewrite the code in such a way that would make sure it is impossible?

Reply:
About 99% of this code was taken directly from Node.js where it's remained largely unchanged for years. It's pretty bulletproof at this point. I'd rather not change it too much here initially.

Reply:
I see. Agree let's keep it close to the original. I'm going to fuzz it anyway :)

Reply:
I did realize that the `KJ_ASSERT`s really ought to be JSG_REQUIREs with meaningful JS errors in case someone decided to hack around with the internal state of the `StringDecoder`. Adding an additional tweak and some additional tests.